src/System.Net.Sockets/src/System/Net/Sockets/SocketPal.Windows.cs

```diff
@@ -116,16 +116,19 @@ public static SocketError Connect(SafeCloseSocket handle, byte[] peerAddress, in
                 IntPtr.Zero);
             return errorCode == SocketError.SocketError ? GetLastSocketError() : SocketError.Success;
         }
-
         public static SocketError Send(SafeCloseSocket handle, IList<ArraySegment<byte>> buffers, SocketFlags socketFlags, out int bytesTransferred)
         {
+            const int StackThreshold = 80; // arbitrary limit to avoid too much space on stack
             int count = buffers.Count;
-            WSABuffer[] WSABuffers = new WSABuffer[count];
-            GCHandle[] objectsToPin = null;
+            Span<WSABuffer> WSABuffers = count < StackThreshold ? stackalloc WSABuffer[count] : new WSABuffer[count];
+            Span<GCHandle> objectsToPin = count < StackThreshold ? stackalloc GCHandle[count] : new GCHandle[count];
+            if (count < StackThreshold)
+            {
+                objectsToPin.Clear(); // important if it is stackalloc, otherwise contents are dangerous garbage that might be accessed in finally
+            }
```

Comment:
Rather than having two checks for the same condition (I don't know if the JIT will eliminate one), why not instead do something like:
```C#
Span<GCHandle> objectsToPin;
if (count <= StackThreshold)
{
    objectsToPin = stackalloc GCHandle[StackThreshold];
    objectsToPin.Clear();
}
else
{
    objectsToPin = new GCHandle[count];
}
```
?

Reply:
the first time I tried that, I got a compilation error from the escape analysis of the stackalloc - will retry

Reply:
![image](https://user-images.githubusercontent.com/17328/41537800-69054246-7301-11e8-8818-543b2ad15012.png)

I'll change it to do the test once only, at least

Reply:
> A result of a stackalloc expression of type...

Yeah, that's annoying... you can work around that by initializing the local to `stackalloc GCHandle[0]`.

Reply:
nice tip; I'd tried `= default`, but there's no way a zero-length `stackalloc` would have occurred to me